ENH: Add var_weights to GLM

thequackdaddy opened the pull request:
Hello,

xref #3371 

This (should) get `var_weights` to work for GLM. I added a Tweedie usecase against R. 

I can't seem to get the Poisson with `var_weights` to `HC0` test that was added (but disabled because of the lack of functionality) to work. The test is here:

https://github.com/thequackdaddy/statsmodels/blob/var_weight/statsmodels/genmod/tests/test_glm_weights.py#L142

It fails on the following `assert`

`assert_allclose(res1.bse, corr_fact * res2.bse, atol= 1e-6, rtol=2e-6)`

Its relatively close... consistently off by a factor of `0.98574`... `corr_fact` brings it much closer... I'm wondering if another adjustment is necessary? 

Honestly, I don't really understand much about sandwiches (except PBJ, meatball, and bánh mì). 

Rest of the test seem to work pretty well. 

I'd be happy to have this merged relatively soon, so thanks for the feedback and review!

josef-pkt commented:
@thequackdaddy Thanks, I look at it soon (as soon as I can ignore 0.8)

If the bse is off by a proportional factor, then there is just a factor to be used to correct this. In some cases I changed our code to match Stata's correction factors, in other cases I adjusted the unit tests so we are more consistent across models than Stata is. I usually reverse engineered the correction factor from the Stata test case with the help, as far as possible, of the Stata documentation. 
(Stata uses different correction factors in GLM than in OLS, while we use the same across all models. more precisely: Stata uses some small sample correction factors in OLS but not in GLM. We match sandwich behavior in GLM to the one of OLS.)

thequackdaddy commented:
Update: There is a bug in the tests... the resids aren't being run (and are wrong). 

I'll fix tomorrow.

thequackdaddy commented:
Turned out to be a much simpler fix than I thought, so I went ahead and did it. Note that the `TestGlmPoissonAwHc` tests lack resids, so those weren't checked in the first go-around of this. 

 R multiplies the deviance residuals and the pearson residuals by the weights (but doesn't for working or response). I'm unclear whether anscombe be multiplied. (I'd be curious if stata also multiples variance weights by these residuals as well...)

I recall that we went back on forth whether resids should be multiplied by frequency weights... but I think frequency weights and variance weights are different. It makes some sense to me that the variance weights should be included in resids, while not so for frequency (as frequency weights is essentially a shortcut to duplicating records)... 

Anyways, I'll pipe down while 0.8 is finalized. That a far bigger milestone.

josef-pkt commented:
about resids: var_weights differ from the other weights because they go inside the likelihood function while the others all just multiply it.

resid_deviance and resid_pearson need the var_weights, 
anscombe would also need the var_weights but I didn't check how easy it is to add them.
resid_response are the raw residuals and should not have the weights.
resid_working are mainly internal for now and might need refactoring (AFAIK, they should match the weights used in IRLS but are currently not used)

thequackdaddy commented:
Thanks for the insights. 

So I'm pretty close then. Right now, I'm including `var_weights` in `resid_deviance` and `resid_pearson` and excluding them from `resid_working` and `resid_response`. I need to include them in `resid_anscombe`, then.

> about resids: var_weights differ from the other weights because they go inside the likelihood function while the others all just multiply it.

This makes sense to me conceptually. I don't know if we have to worry about this practically, though. 

For all the families, it appears that `var_weights` is associative--you can apply it inside the likelihood/deviance functions or outside. 

So my shortcut was just to multiply `np.sqrt(var_weights) * resids` like I did [here](https://github.com/statsmodels/statsmodels/blob/1a01699b5a394afa0952d676391b237d38052868/statsmodels/genmod/generalized_linear_model.py#L1338) and [here](https://github.com/statsmodels/statsmodels/blob/1a01699b5a394afa0952d676391b237d38052868/statsmodels/genmod/generalized_linear_model.py#L1338). I'd assume that I could do to same [here](https://github.com/statsmodels/statsmodels/blob/1a01699b5a394afa0952d676391b237d38052868/statsmodels/genmod/generalized_linear_model.py#L1351) for anscombe, if necessary.

However, if you try to calculate the deviance from the respective methods in `family`, this shortcut won't support that. Not sure if we need that? Adding `var_weight` as a parameter to a all the `resid` methods in `family` should be easy enough... I was trying to be lazy. 

@josef-pkt Would it be possible (easy) to add the resids to for TestGlmPoissonAwHc [here](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/tests/results/results_glm_poisson_weights.py#L1157). Otherwise, I won't have a unittest to compare against anscombe (I don't know how to get anscombe residuals from R's glm.)

josef-pkt commented:
I just ran the Stata file again to check

residuals, parameter estimates and similar are independent of the cov_type. IIRC that was the reason not to include the same residuals several times

```
>>> results_poisson_aweight_nonrobust.resids - results_poisson_aweight_hc1.resids
array([[ 0.,  0.,  0.,  0.,  0.,  0.],
....
```

To avoid having to adjust the tests, you could add at the end of the results module, 
the colnames and rownames might not be needed

```
################  not autogenerated
results_poisson_aweight_hc1.resids = results_poisson_aweight_nonrobust.resids
results_poisson_aweight_hc1.resids_colnames = results_poisson_aweight_nonrobust.resids_colnames
results_poisson_aweight_hc1.resids_rownames = results_poisson_aweight_nonrobust.resids_rownames
```


related asides: 
- `hat` or `predict_hat` in `predicted` is not available after robust cov_types in Stata (including when pweights are used)
- the `est` results are not in the same order when I run the Stata `do` file again, that's irrelevant for the code but messes up the file comparison between new and old version

thequackdaddy commented:
Oh duh. I hadn't looked at those, and now it shows an inconsistency. 

Based on that, stata does *not* utilize `aweights` (or `fweights`) in any of the resids.

```python
In [4]: from statsmodels.genmod.tests.results import results_glm_poisson_weight
   ...: s as res

In [5]: x = res.results_poisson_aweight_nonrobust.resids

In [6]: y = res.results_poisson_fweight_nonrobust.resids

In [7]: from numpy.testing import assert_allclose

In [8]: assert_allclose(x, y)

In [9]:
```

Options: 
* Follow R - Include `var_weights` in `resid_deviance`, `resid_pearson`, and presumably `resid_anscombe`.
* Follow stata - Don't include `var_weights`

thequackdaddy commented:
And to recap: the only real differences I see so far in `freq_weights` and `var_weights` is:
* `freq_weights` alters the degrees of freedom of the residual (impact on `scale` mainly)
* `var_weights` (presumably) is incorporated in some of the `resids`. 

Other than that, they appear to be functionally the same.

josef-pkt commented:
We follow R in this.  At least in my recent thinking that is the more obvious use case for var_weights. i.e. those resids are approximately standard normal without heteroscedasticity.
(Whether it's the best for other use cases like producing sandwich covariances or hypothesis testing is a different issue, and we can work around or even change again later.)

Binomial variance function includes `n`, and so also those resids, which would be a similar normalization as var_weights.

josef-pkt commented:
Based on what I remember from the SPSS documentation, they have a few places where freq_weights and var_weights are handled differently.

Also, given the recent discussion, we need unit tests with noncanonical links. Does tweedie cover that enough?
Gamma and Gaussian with log-link might be useful candidates. Probit is more common, but I ran into some precision problems comparing GLM-Binomial-Probit and discrete.Probit in the Hessian that I couldn't figure out. NegativeBinomial defaults to a non-canonical link but uses dispersion.

thequackdaddy commented:
> Based on what I remember from the SPSS documentation, they have a few places where freq_weights and var_weights are handled differently.

Cool. If you stumble across that again, it would be good to spot check that we are doing it the same way. 

> Also, given the recent discussion, we need unit tests with noncanonical links. Does tweedie cover that enough?

I'm using a non-canonical link at the moment (var_power is 1.55 in the model, so the canonical link is `mu ** -0.55 / -0.55` but I'm using `log(mu)`). In my larger datasets (5.4 million records) I'm getting MUCH closer matching `(res.mu * model.var_weights).sum() == (model.endog * model.var_weights).sum()` within 1%. Still off by a tad... not sure I really care (or should care?). The hessian is positive definite. 

I have an R script that generated the Tweedie result. Should be trivial to get others tests incorporated. I'll do gamma and gaussian with log link. If you have any others, let me know.

thequackdaddy commented:
I am struggling to remember something...

For gamma/gaussian, `loglike != -2 * deviance`, while R sort of assumes that it does. 

I feel like I studied the reason for this earlier, but can't find it at the moment.

Besides that, the tests, look good. They will fail here, however, because of this limitation.

josef-pkt commented:
no, deviance = - 2 (loglike(model) - loglike(saturated model))
I'm not sure I remember the sign correctly
I just used this in a unit test for fixing poisson deviance

saturated model has `mu = endog`  (with bounds handling)

(It's like a likelihood ratio test statistic against the saturated model, but it's not valid as standard LR test unless there are only a fixed finite number of unique exog combinations. AFAIR)

thequackdaddy commented:
Thanks for the comments. 

I think the issue is that loglikelihood appears to be different for gamma/gaussian between R/stata/statsmodels. 

Here's some R code:

```R
> data <- read.csv("c:/users/pquackenbush/git/statsmodels/statsmodels/datasets/cpunish/cpunish.csv")
> data$INCOME <- data$INCOME / 1000
> model <- glm(EXECUTIONS ~ INCOME + SOUTH, data = data, family = Gamma(link = "log"))
> summary(model)

Call:
glm(formula = EXECUTIONS ~ INCOME + SOUTH, family = Gamma(link = "log"), 
    data = data)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.5203  -0.7113  -0.4504  -0.0993   2.0244  

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept) -0.52963    2.63861  -0.201   0.8438  
INCOME       0.02915    0.07053   0.413   0.6857  
SOUTH        1.61720    0.68678   2.355   0.0337 *
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for Gamma family taken to be 1.671754)

    Null deviance: 25.680  on 16  degrees of freedom
Residual deviance: 15.449  on 14  degrees of freedom
AIC: 81.462

Number of Fisher Scoring iterations: 8

> logLik(model)
'log Lik.' -36.73083 (df=4)
```

And here's the same thing in statsmodels:

```python
In [1]: import numpy as np

In [2]: import statsmodels.api as sm

In [3]: import statsmodels.formula.api as smf

In [4]: data = sm.datasets.cpunish.load_pandas().exog

In [5]: data['EXECUTIONS'] = sm.datasets.cpunish.load_pandas().endog

In [6]: data['INCOME'] /= 1000

In [7]: model = smf.glm('EXECUTIONS ~ INCOME + SOUTH', family=sm.families.Gamma
   ...: (link=sm.families.links.log()), data=data)

In [8]: result = model.fit(atol=0, rtol=1e-25)

In [10]: result.llf
Out[10]: -38.99740362821273

In [12]: result.aic
Out[12]: 83.99480725642546
```
All the `param`, `cov_params`, and `resids` and `deviance` appear to be the same between each. 

(Notice AIC and loglike are different)

There's some commentary about this in the [test file here as well](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/tests/test_glm.py#L118)...

As for gaussian, I would guess there are inconsistencies as well, although its not noted. Notice we use a [very relaxed match tolerance](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/tests/test_glm.py#L263).

R appears to really use `dgamma(endog, sum(var_weight) / scale, scale=mu * scale / sum(var_weight), log=TRUE) * var_weight`. I'm thinking I'll have to do some algebra to incorporate this into our loglike function.

thequackdaddy commented:
Bookmarking this stata manual for reference, pages 31-32:

http://www.stata.com/manuals13/rglm.pdf

thequackdaddy commented:
OK this is why I've come to get frustrated with R. They have all sorts of weird algebra going on that is rarely (read, never) well documented. 

R saves an AIC score in its model objects, and loglikelihoods are backed into using the saved AIC score. For gamma, its relatively straightforward once you re-state the gamma pdf in exponential form. (Use 1 / scale for gamma's shape and mu * scale for gamma's scale.) Additionally, `scale` is calculated as `deviance / nobs` for these 2 families, while statsmodels (and presumably stata) use `deviance / df_resid`. For gaussian, they've identified a bunch of constants in the summation and simply removed them and multiplied by the number of records (nobs) presumably to make the computation cheaper. Seems a little overkill because it makes the readability really poor. In my mind, it would be far, far easier to just use `dgamma`, `dNormal` etc. as that is far more readable (in my mind) and shouldn't be a much more expensive computation. 

So I'm leaning towards thinking loglike is somewhat "special" compared to the other metrics we have. I have some questions on how we proceed then...

* stata uses a very simple formula (who's documentation I cannot find) for loglike weights. Essentially, its `weights = freq_weights` when `freq_weights` are specified and `weights = nobs * var_weights / sum(var_weights)` when `var_weights` are specified. Then just multiply these values by the basic loglike function. The stata manual is very helpful for this. Note that based on the language, it sounds like stata only supports `freq_weights` or `var_weights`, and not both. 
* R doesn't really have support for `freq_weights`... and the designers spent a good amount of time re-writing and simplifying their `aid` (and hence, `loglike`) functions to minimize calculations.  
* For both tools, it appears deviance is calculated the same and is simply `sum(iweights * resid_dev)` and there are no noticeable (as far as I see) differences between them

I suggest the following... 
* For loglike, follow stata more than R. 
* I *think* stata limits to either `var_weights` or `freq_weights` using their formula. I'm wondering if we could get the same effect with the formula `sum(freq_weights) * iweights / sum(iweights)` (recall `iweights = freq_weights * var_weights`

josef-pkt commented:
@thequackdaddy   I see you are having fun :)
It reminds me of what Skipper was complaining about during the initial rewrite of GLM and comparison to R and other packages.

One problem is that open source packages don't have the army of professional documentation writers like Stata, SAS and SPSS. R is ok when there is a full description of the methods and formulas in a book or vignette, but the documentation itself is usually not very complete. Neither is the one of statsmodels.

However, in general we stick much closer to text book definitions even giving up some computational shortcuts. For loglike, we always use the full loglikelihood function and not dropping any terms that are constant during optimization. For aic and bic we inherited some inconsistent definitions #1733
SPSS manual has explicit formulas for both the full loglike and the loglike used during optimization which drops the terms that are irrelevant when estimating the mean function parameters.

Also, after figuring out what other packages are doing, we still have to decide what we want to do.
E.g. Stata uses the normalized weights for most things, which we started to avoid for variance weights.
#1170 

(I expect that it will still take some time before I work through all the details here, and try to figure out how to adjust the cov_type sandwiches.)

thequackdaddy commented:
Sounds good. I'll eagerly await your review/insights.

josef-pkt commented:
You can also skip some of the test for those extra results if it's difficult to work around conflicting definitions or shortcut.

E,g. one test for loglike would be that irls which doesn't use the loglike and the scipy optimizers which do use loglike converge to the same parameter estimates. Gaussian can also be compared with WLS. If those work, then it means we have, at least, the main results correct.

(I'm commenting without having looked at your code yet.)

josef-pkt commented:
another one: 
information criteria like aic and bic across different definitions/shortcuts should only differ in an additive constant, so ranking is preserved (the correction term is the same for different numbers of explanatory variables).
One extra possible variation is whether aic, ... are normalized by nobs or not, i.e. differ by a factor that is independent of the number of explanatory variables.
(In general, we default to full llf in the information criteria, but there are variations across packages on dropping items that are irrelevant for ranking alternative models within one model class with the same endog.)

thequackdaddy commented:
bookmarking this:

http://www.cpc.unc.edu/research/tools/data_analysis/statatutorial/sample_surveys/weight_syntax

I tried to use `fit(method='bfgs')` to check the loglikelihoods, and it flat didn't work. Need to re-think some of this. I think I need to understand how the loglikelihood would change with different weight assumptions. I don't understand them all, to be frank. 

Trying to get my head around pweights/iweights vs the rest. (This PR might be a decent opportunity to include them). 

* pweights - I'm not at all familiar with survey data but from what I can tell if I have a dataset with 10 records (observations) and these observations represent 10 different groups/samples, pweights would be used to assign higher credibility to the groups. I'm not sure of the mechanics here. 
* fweights - Frequency weights. How many times the same observations (endog/exog inclusive) occurs in data. The mechanics here are that the `scale` parameter will be updating (assuming that freq_weights are integers >= 1, I would expect it to increase `df_resid` and lower `scale`. This will be used to calculate the `bse` and `cov_params`. 
* aweights - Analytic (variance) weights. Used for modeling on mean (and rate?) data. A larger weight implies that increased sampling accuracy. I presume the difference here is that a larger pweight might really mean unchanged (or lower?) sampling accuracy because a larger pweight means that the observation reflects a larger, potentially more heterogeneous group. I had thought they work by scaling loglikelihoods, but its not just `var_weight * loglike` or else stata is wrong (doubtful). 
* iweights - user-defined, so I think they can do whatever you want them to do. Not sure we need to include them? 

It seems reasonable (to me) to assume that weights should all result in (nearly) the same params, the difference will be `scale`, `cov_params`, `resids` and the descendants thereof. (Not totally sure about `pweights`.)

josef-pkt commented:
I thought of iweights as every other kind of weights with no guarantee that the auxiliary results have a "weight" interpretation. E.g. when we are mainly interested in the parameter estimates for prediction, which is the same for every interpretation of weights, at least in the basic GLM models.

pweights (prob_weights) work for most parts like freq_weights. The main difference will come from the implied nobs, e.g. for the computation of the scale.
For survey and similar weighting, the p_weights adjust the weights from the data (distribution of the exog) to some "fictitious" population (with a different distribution of the exog).

The problem for the implementation and a full weights-PR is that in each case we have to figure out and decide how to handle the additional results, like scale, degrees of freedom, effective nobs and it's implication for cov_params. 

For the basic computation, I think we have only two types of weights, var_weights and all others.

thequackdaddy commented:
Sounds good. On first glance, the tests are really failing because the data I was using (cpunish) isn't well modeled by either gaussian/gamma. There's some positive definiteness problems and other issues that will need to be sorted out. 

Additionally, the robust residuals don't seem to work. I stumbled across this paper: http://www.parisschoolofeconomics.eu/docs/dupraz-yannick/using-weights-in-stata(1).pdf

There's a formula on top of page 5 that I'm wondering if we need to use?

josef-pkt commented:
What do you mean by "robust residuals"?  some residuals or robust cov_type

cov_type != nonrobust might not be correctly wired up or need to be adjusted for weights. The dupraz-yannik description is for linear model. We should have the equivalent GLM version when everything is wired up correctly, e.g. weights in the scale computation with weighted resid_pearson or score. wls_results.normalized_cov_params should have the weighted inv(x'x) part.
bfgs or similar needs score/score_factor and hessian/hessian_factor to be adjusted, I think.

thequackdaddy commented:
@josef-pkt You'll have to forgive me on my last message. I was doing something dumb and was tired and wrote something non-sensical. I don't even know what I meant looking at the results. 

I've managed to get a bunch of comparisons of `irls` vs. `bfgs`/`newton`. I chose `bfgs` when possible as I seem to recall from my notes that you recommended it at one point. I also believe that `bfgs` uses the `loglike` function (and `hessian`/`score`) while `newton` only uses `hessian`/`score` for the optimization (and `loglike` only for reporting. 

I started with the irls vs. gradient code in the `test_glm` code. I applied higher `var_weights` on obsevervations  where `endog > endog.mean()` so that larger values would get more weight. I figured this provide a test that weights are actually doing something--i.e., regressing towards the larger values. I think this intentional bias might be causing a few of the additional convergence issues I've run into. 

So here's what I have. Please have a look and let me know what you think. Happy to just apply weights randomly, but I wasn't confident that would account for much.

josef-pkt commented:
I just remembered the aggregation issue that I tried out for Tweedie in a notebook. and #2840

I think we should have a generic equivalence between repeated, aggregated and averaged datasets, similar to what we used for freq_weights.

- repeated - individual observations where some have the same exog but different endog -> GLM
- aggregated - we use unique exog, but aggregate the endog by samples with same exog -> GLM with exposure
- averaged - we use the unique exog, but average the endog by samples with same exog. -> GLM with var_weights

Those should give identical final results for cov_type='nonrobust' but most likely not the same cov_params/bse for any of the robust cov_types. (*)
I think we should be able to use this also to get exposure to work "correctly".

I'm almost (!) sure that applies also to non-canonical link functions.

(*) I still have problems fitting overdispersion or, more generally, heteroscedasticity in there.
AFAICS, this aggregation, averaging, using exposure and `aweights` for nobs in average only works if we assume that the distribution is correctly specified. If we want to allow for overdispersion, or heteroscedasticity or correlation as in the cov_types, then aggregation and averaging will break down, maybe unless we assume that there is no heterogeneity within the subsamples that we aggregate or average, but could be allowed between the aggregated or averaged subsamples.

in your case this would mean, I think: Aggregating over time for individual policies assumes that the underlying distribution is unchanged or homogeneous, but nonrobust cov_types could allow for correlation or dispersion/heteroscedasticity across policies.

(I'm still struggling with what is allowed or assumed if we allow for misspecification.)

Aside: I read the Dupraz tutorial/explanation a few days ago. It's a good introduction for the ideas behind the weights but it is misleading for the computation that we need. He incorporates the sqrt(weights) in the exog, similar to wexog in WLS. However, this only works for quadratic terms which are the only terms he is using wexog.T @ wexog, wexog.T @ wendog  wresid.T @ wresid.
This does not work if we want a weighted mean for example 
`(weights * endog).sum() != wendog.sum() = (sqrt(weights) * endog).sum()`
(This was one of my confusions in fixing cov_type for freq_weights.)

(I still think that weights is the most difficult to understand among all the generic topics.)

thequackdaddy commented:
> (I still think that weights is the most difficult to understand among all the generic topics.)

Prior to working on this, I always thought this was a relatively straightforward topic. Its definitely a topic that has exploded well beyond what I thought. 

I went ahead and made some comparisons of repeated vs. aggregated and repeated vs. average. `params` and `bse` match, but `deviance`, `loglike` and `resids` don't. I *think* all that makes sense. I'm only using Poisson with log and non-robust cov_type. Perhaps I should try something more exciting (maybe Tweedie 1.5 with log?) to test non-canonical links.

josef-pkt commented:
It's useful to have the basic models working so we don't have to think about the influence of additional "complications". Given that tweedie is your main use case, I think you should add those tests. Aggregate/exposure will not work yet, because we need to change the interpretation and implementation of exposure for non-canonical links, or at least for tweedie dispersion.

josef-pkt commented:
Related #3079  I still haven't looked at it carefully. Is the exposure part related to the exposure problems that we were discussing now?

(However, it's good to understand the background, but this PR is for var_weights.)

thequackdaddy commented:
Ok. I'll try to add Tweedie in the coming days/hours. 

I tried `cov_type=HC0` and sure enough, it failed. I'd be interested to incorporate it if you have an idea of how to do it... or maybe a "robust covariance for dummies" reading. The idea of having standard errors that are robust to heteroskedasticity is obviously appealing... but I don't see how weights would fit in to the sandwich. 

As for #3079, I was holding off on it. I still think our `estimate_tweedie_power` is not robust at-all.. mostly doesn't work for me. But I was thinking it would also be very helpful to include tweedie's `pdf`/`cdf` in either in here or `statsmodels.distributions`. The R `tweedie` package can do either pdf's or cdf's... so you can do qq plots and the like. I'd like to get that too... eventually. 

Like you said a while back... there is no bottom to the ocean.

josef-pkt commented on statsmodels/genmod/tests/results/res_R_var_weight.py:
```diff
@@ -0,0 +1,8615 @@
+res['ll'] = np.nan 
+res['resids_colnames'] = ['resid_response', 'resid_pearson', 'resid_deviance', 'resid_working'] 
+res['resids'] = np.array([-0.51916916931427282, 2.5547051833524539, 0.40292815811196381,
+ 0.27080945814608781, 3.9285319711191033, 3.9285319711191033,
```
if you have large arrays, then it would be better to save them as csv files and load them from there

The main reason is that counting lines of code can get very biased with python modules like this. A not merged PR added 100,000 or so lines of python code almost all from test data, which I thought is too much given that currently we are a bit above 150,000 LOC of python.

thequackdaddy replied:
OK. Initially it was only 20 lines because there were no line breaks. 

I'll make a csv.

josef-pkt commented on statsmodels/genmod/tests/test_glm_weights.py:
```diff
@@ -182,3 +221,367 @@ def setupClass(cls):
+
+class TestGlmTweedieAwNr(CheckWeight):
+    @classmethod
+    def setupClass(cls):
```
rename to `setup_class`
we will switch to pytest in future which doesn't support the camelCase

josef-pkt commented on statsmodels/genmod/tests/test_glm_weights.py:
```diff
@@ -182,3 +221,367 @@ def setupClass(cls):
+
+    skip_one = False
+    for family_class, family_links in families:
+        for link in family_links:
```
Is this part mostly copied from another test?

(just a question for now, but we might want to eventually reuse code.)

thequackdaddy replied:
Yes.

josef-pkt commented:
I hate the new github review interface. I never know where I am, nor where to find the comments again.

josef-pkt commented:
> I tried cov_type=HC0 and sure enough, it failed.

I think this is not a "failure". AFAIU without having it verified yet, this is a "feature".  Disaggregated data allows heteroscedasticity for all observations but no correlation. Averaged or aggregated data allows for a different kind of heteroscedasticity and maybe also for some within aggregate correlation.

thequackdaddy commented:
In the most recent commit I:
- Added a comparison for `Tweedie(var_power=1.5, link=log())` for repeated vs. aggregated and repeated vs. average
  - repeated vs. aggregated seems to not match to precision. I've increased the param precision enough that it should match, and yet it doesn't. My guess is this has to do with the problems identified earlier. (Its really close, but both should have converged to precision, I think
  - repeated vs. average doesn't match on the `bse`... but it off by a consitent factor of `0.93...`. I'm confused why this would be. (Perhaps `scale` needs an adjustment?)
- Changed the really long .py file to a .csv. Note that this doesn't reduce the lines of code in the commit.
- setupClass -> setup_class (Note I only did this for the added tests... I think bashtage's commit already takes care of this so I don't think I want a merge conflict by doing changes outside this scope)
- A couple of other small things that I noticed.

thequackdaddy commented:
@josef-pkt I had some success with this latest commit. I realized how to get the aggregate/exposure to match (still have to use var_weights) and realized that averaged won't work because the scales are different (doesn't apply to Poisson because scale is assumed to be 1).

So I think I'd appreciate any review that can occur.

I'll be offline next few days. (I'm going to Ottowa) but I'll respond after I return.

thequackdaddy commented:
@josef-pkt Just pinging to see if you think you'd could provide some feedback on this. 

Thanks for your help. I appreciate it.

josef-pkt commented:
I briefly read/skimmed through the code, but I don't see any problems :(
That means I need to get back into some of the details to remember what is going on.

One idea to get an overview is to compare equivalent models and check in a example script or notebook which results attributes differ across versions.
for example I did this with absorbed exog in OLS
https://github.com/statsmodels/statsmodels/pull/2774#issuecomment-173362025

Any non-matching results need a justification or, if there is none, a warning in the docstring.
(I haven't checked yet how comprehensive the unit tests against Stata are, which usually includes many of the result statistics.)

josef-pkt commented:
unit tests with respect to Stata in `CheckWeight`: `params`, `corrfact * bse`, `llf`, `deviance`, and 4 types of residuals (plus unlikely resid_anscombe, IIRC Stata doesn't have it)
sounds good, maybe unclear if we have df_resid (in case we use `use_t=True`)

cases with aweights: TestGlmPoissonAwNr, TestGlmPoissonAwHC, 
from before with freq_weights: TestGlmPoissonFwNr, TestGlmPoissonFwHC, TestGlmPoissonFwClu, 

(cov_type code Nr: nonrobust, HC: HCx only one, HC0 or HC1 is available, Clu: cluster 1-group)

against R: same attributes (CheckWeight) as with Stata, except different definition of `llf`
TestGlmGaussianAwNr, TestGlmGammaAwNr, TestGlmTweedieAwNr 

``` python 3.6 travis
statsmodels.genmod.tests.test_glm_weights.TestGlmGammaAwNr.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmGammaAwNr.test_r_llf ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmGammaAwNr.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmGaussianAwNr.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmGaussianAwNr.test_r_llf ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmGaussianAwNr.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonAwHC.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonAwHC.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonAwNr.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonAwNr.test_residuals ... ok
/home/travis/miniconda2/envs/statsmodels-test/lib/python3.6/site-packages/statsmodels-0.8.0-py3.6-linux-x86_64.egg/statsmodels/genmod/generalized_linear_model.py:1326: SpecificationWarning: cov_type not fully supported with freq_weights
  SpecificationWarning)
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonFwClu.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonFwClu.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonFwHC.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonFwHC.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonFwNr.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonFwNr.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonPlain.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonPlain.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmTweedieAwNr.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestGlmTweedieAwNr.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestRepeatedvsAggregated.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestRepeatedvsAggregated.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestRepeatedvsAverage.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestRepeatedvsAverage.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestTweedieRepeatedvsAggregated.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestTweedieRepeatedvsAggregated.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.TestTweedieRepeatedvsAverage.test_basic ... ok
statsmodels.genmod.tests.test_glm_weights.TestTweedieRepeatedvsAverage.test_residuals ... ok
statsmodels.genmod.tests.test_glm_weights.test_wtd_gradient_irls ... ok
```

https://coveralls.io/builds/11062371/source?filename=%2Fhome%2Ftravis%2Fminiconda2%2Fenvs%2Fstatsmodels-test%2Flib%2Fpython3.6%2Fsite-packages%2Fstatsmodels-0.8.0-py3.6-linux-x86_64.egg%2Fstatsmodels%2Fgenmod%2Ftests%2Ftest_glm_weights.py

josef-pkt commented:
Based on the test coverage the core results look good against Stata and R. Then I don't see a reason not to merge this soon. We already have a warning that weights are not supported with all options, and we could leave the missing support to follow-ups. (*)

Two python versions on Travis got killed in these tests. I doesn't specify a reason.
@thequackdaddy  I think you could rebase this again on master and we can check travis to see if the problems persist.

(*) Based on the Stata test `cov_type='HC1'` works, other cov_types still need to be checked. Offset should be fine, even if we don't have test coverage. `exposure` will need work outside of Poisson with log link.
I haven't tried again to figure out the average/aggregation/... equivalence.

josef-pkt commented:
API and internal use of `iweights` looks good, but I'm not ruling out that we need another refactoring before we are finished implementing "all kinds of weights".

thequackdaddy commented:
So the travis builds fail... 

I think what is happening is that between these 2 tests:
```
statsmodels.genmod.tests.test_glm_weights.TestTweedieRepeatedvsAverage.test_residuals
statsmodels.genmod.tests.test_glm_weights.test_wtd_gradient_irls
```

Travis seems to hangs. I can reproduce this on my local windows laptop using a python 2.7 environment. The key... and this I do not understand... is that when I run `sm.test(label="full", verbose=3)`, it takes a nap. It wakes up about 60 seconds later and runs the next test just fine on the laptop, but I guess that nap is enough time for Travis to think it died.

josef-pkt commented:
Is the CPU busy during the "nap", or is it just doing nothing for this time? 
Does it also happen if you just run the specific test module 
`nosetests statsmodels.genmod.tests.test_glm_weights`  or path of file

thequackdaddy commented:
Wait (as usual) was looking at the wrong thing. Of of the tests runs really quickly on python 3.4, 3.5, 3.6, but is slow on 2.7 and 3.3 because I generate a large random matrix. I shrank the matrix and it seems to work now.

josef-pkt commented:
We need to check Binomial to see if the combination of counts larger than 1 and var_weight works. Binomial is an outlier in the general pattern and is not necessarily covered if everything is correct generically.
In standard usecase specifying var_weights and number of (successes, failures) will be substitutes, but not for (over) dispersed Binomial (#2632)

thequackdaddy commented:
I made a notebook trying to show off the 3 different weight cases... 

`repeated` -> `freq_weights`
`averaged` -> `var_weights`
`aggregated` -> `exposure`

(I forgot to clean the notebook so I just amended the commit. Sorry for the extra push!)

In addition, I added a test for binomial using the 1-dimensional array `endog` for binomial. I'm unclear how (or if we should) we can use 2-dimensional array `endog` for binamial along with weights... So I didn't add a test for that.

thequackdaddy commented:
@josef-pkt I re-read your note about `binomial` and think I understand what you want to check... and I think you uncovered something. 

Here's a notebook (scroll to the bottom) where I tried `var_weights` and comparing to a 2-dimensional endog.

https://gist.github.com/thequackdaddy/a483d7d54bd0cb8c773936082392c7ce

So to confirm... 

A 2 dimensional endog with each row representing number of successes/failures gives us `n_trials` of `successes + failures`. That should give the same parameters, standard errors, likelihood, and deviance as  `endog = successes / (successes + failures)` with `var_weight = successes + failures`, correct?

Additionally, can you outline a type of test you'd like to work with a 2-dimensional endog with `var_weights`? I have a hard time visualizing overdispersed binomial...

josef-pkt commented:
@thequackdaddy  I will try to look tonight. It's mostly family weekend (live butterfly exhibit is next :).

overdispersed binomial: It's similar to aggregation issue we discussed before. If one observation is for example the counts for a group/cluster, then it could be that the success probabilities are correlated for individuals within the groups. However, binomial assumes independent trials and would have the wrong variance.

josef-pkt commented:
The notebook in the gist looks great.
params and wald inference seems to agree everywhere, but I need to think about the differences in llf, deviance and pearson chi2. 
My impression is that likelihood ratio test should give the same answer in equivalent representation. But I'm not sure yet whether there is a scaling that is missing or whether it reflects the transformation of the endog.

thequackdaddy commented:
Sounds good. I hope the butterflies are fun. 

I got to spend the rest of my afternoon working on income tax. I filed last year and so I naturally assume I was good. But now I find out you have to file every year. 

I think crawfish will make me happy.

josef-pkt commented:
We have a problem with #3543
there will be a large number of merge conflicts  
there are currently bugs in scale handling in master

to the first
based on the changes, it looks like it would be better to merge #3543 first because that has many individual changes, while this PR has mainly the substitution freq_weights to iweights in family.py

to the second
scale for under- and overdispersion is not used in this PR. However, I got suspicious and remembered the scale issues because var_weights should follow a similar pattern in the code as scale, which it doesn't because of the scale bugs.

I'm still not sure about the loglike and similar differences:
e.g. pearson_chi2 computation looks correct to me, but it doesn't take into account that the endog is transformed, for example when using binomial with proportion instead of counts. I don't think we are supposed to take the endog transformation into account given that the user is doing it and not statsmodels. We might have to specifically check what R and Stata are doing in the var_weights case if it is used for averaging.

josef-pkt commented:
If var_weights are interpreted as QMLE (quasi instead of correctly specified MLE) in families with fixed scale=1, then the inference based on loglike and deviance would not apply, similar to the cov_type != 'nonrobust' case, and we wouldn't need to match llf and deviance.
However, in the "average" transformation case we could compensate for the averaging to get a valid loglikelihood similar to the internal binomial transformation.

josef-pkt commented:
AFAICS, pearson_chi2 has currently no test coverage.
Based on my derivation is should be the same with var_weights and counts. But to me it looks like the current count version does not include `self._n_trials` which looks like a bug. resid_pearson contains n_trials. Note: binomial endog is rescaled to proportion, AFAICS based on the code

thequackdaddy commented:
Thanks for the review. 

I think I know how to fix the following issues... 
* Write tests to compare `pearson_chi2` to R (I think `sum(residuals(model, 'pearson'))` should work
* Include `var_weights` is binomial family to work in lieu of count (num successes + num fails) (I'm still confused how--or if--you could have a 2 dimensional `endog` for binomial, and use `var_weights`. (This should make deviance, likelihood, and pearson all match.
* Include `self._n_trial` in pearson

Things you still want to investigate: 
* Should `var_weights` on an "averaged" dataset result in the same likelihood, deviance, pearson as an "unaveraged" dataset.
* How or if scale/dispersion parameter should be accounted for

Things I have to wait for:
* This merge conflicts with the 3543. After that's merged, deal with the merge conflicts. 

That sound right? I can probably get those first 3 things done... they seem easy enough.

josef-pkt commented:
I'm not sure about the second item

> Include var_weights in binomial family to work in lieu of count

I think R is awful with treating weights differently depending on which family is used. I would like to have a consistent definition or an extra keyword/option whether var_weights are average or QMLE.
Given that we explicitly implement the count version for binomial, I don't think we **have** to have the average interpretation for llf and deviance.
I guess we should be able to fix pearson_chi2.

The Stata results include chi2, but they are currently not use in an assert.

thequackdaddy commented:
Just thinking about API here, then, would you like to have a new keyword (`binom_trials` maybe?) and have `var_weights` just be QMLE?

The idea of `var_weights` meaning `average` for binomial does seem odd to me...

josef-pkt commented:
I feel like leaving the question open for this PR and stick to the current QMLE interpretation for llf with var-weights. (I try to compare the computation with the SPSS formulas.)

If we have an extra keyword now, then we would have to check the behavior for it for all families.

Also some of the usage for var_weights will come from double-exponential or similar dispersion models. But I don't know enough about the computational details to tell which version would be better right now.

Related: I haven't figured out how your Poisson average model works. It's not obvious and I didn't spend enough time on it.

Also: the stata test results have several versions of `deviance`. One possibility would be to add a second version of llf and deviance. But I'm not sure how it would be linked to a generic LR or deviance test if there are more than one version of llf and deviance.

josef-pkt commented:
to confirm the pearson_chi2 bug with binomial in master

last example of gist notebook (I renamed `res` to unique names)

```
(res_binom.resid_pearson**2).sum(), res_binom.pearson_chi2, (res_binom_wt.resid_pearson**2).sum(), res_binom_wt.pearson_chi2
Out[24]:
(4053.3993437537383, 9.5037638756406455, 4053.399343753348, 4053.3993437533463)
I
```

josef-pkt commented:
type(glm_binom_wt.iweights) is pandas Series
I think this should be a numpy array, so we don't have to rely that Series behaves like ndarray

josef-pkt commented:
most likely another bug in family Binomial deviance for this usecase.
AFAICS, deviance should be the same with n_trials or var_weights=n_trials.

However, the branching in binomial.dweviance **assumes** that endog is binary if `self.n == 1`
`one = np.equal(endog, 1)` converts proportions to binary
#3543 removes the branching and uses endog in all cases

This fix should make deviance the same in Binomial-counts and Binomial-var_weights, AFAICS

thequackdaddy commented:
Thanks for you comments, Josef. 

Changes 
* Renamed notebook to glm_weights which is more consistent (the others have underscores, not dashes)
* Added pearson_chi2 tests. Note that this passes for the unweighted version, but fails for the weighted version. I'm not exactly sure what's happening here. I am attempting to compare `res1.pearson_chi2` to `res2.chi2`... They aren't even in the same ballpark, so I don't know if we're doing something really different here. I notice that the values `chi2` numbers from stata appear to be different for different `cov_types`. Does this make sense? Or does `chi2` mean something different? 

```
======================================================================
FAIL: statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonAwHC.test_pearson_
chi2
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Program Files\Anaconda3\envs\patsytest\lib\site-packages\nose\case.py
", line 197, in runTest
    self.test(*self.arg)
  File "C:\Users\pquackenbush\git\statsmodels\statsmodels\genmod\tests\test_glm_
weights.py", line 94, in test_pearson_chi2
    rtol=1e-6)
  File "C:\Program Files\Anaconda3\envs\patsytest\lib\site-packages\numpy\testin
g\utils.py", line 1411, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "C:\Program Files\Anaconda3\envs\patsytest\lib\site-packages\numpy\testin
g\utils.py", line 796, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Not equal to tolerance rtol=1e-06, atol=1e-06

(mismatch 100.0%)
 x: array(15.486301129253283)
 y: array(356.663774965618)

======================================================================
FAIL: statsmodels.genmod.tests.test_glm_weights.TestGlmPoissonAwNr.test_pearson_
chi2
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Program Files\Anaconda3\envs\patsytest\lib\site-packages\nose\case.py
", line 197, in runTest
    self.test(*self.arg)
  File "C:\Users\pquackenbush\git\statsmodels\statsmodels\genmod\tests\test_glm_
weights.py", line 94, in test_pearson_chi2
    rtol=1e-6)
  File "C:\Program Files\Anaconda3\envs\patsytest\lib\site-packages\numpy\testin
g\utils.py", line 1411, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "C:\Program Files\Anaconda3\envs\patsytest\lib\site-packages\numpy\testin
g\utils.py", line 796, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Not equal to tolerance rtol=1e-06, atol=1e-06

(mismatch 100.0%)
 x: array(15.486301129253283)
 y: array(94.62492461274286)
```
* I have not changed deviance/loglike as I think the needed rebase/merge makes this too difficult. 

Let me know what else you think this needs... (or if I missed anything).

Thanks, Josef.

thequackdaddy commented:
This might pass if we can use one of these values... not sure what they all mean, though. 

From the stata result file
```
           deviance_ps = 15.48630027479802,
           deviance_p = 15.48630027479802,
```

Should I compare this to `res1.pearson_chi2`?

josef-pkt commented:
I'm checking the Stata definitions. I didn't remember or never checked the details for this

deviance_p   Pearson deviance  -> this should be our pearson_chi2
deviance_ps  scaled Pearson deviance  - divides by scale

disperse_p  Pearson dispersion ->  pearson_chi2 / df_resid

chi2 is most likely the statistic for the comparison with the null, constant-only, model
In the Stata results, the chi2type is "Wald", while we use likelihood ratio for nonrobust, AFAIR.

josef-pkt commented:
aside: Stata rescales aweights (and pweights) to sum to nobs. In WLS we decided not to do that. So things that depend on the sum of weights but not on the compensating scale will differ between Stata and statsmodels.

thequackdaddy commented:
@josef-pkt Ok. Thanks. Amended the last commit to test `assert_allclose(self.res1.pearson_chi2, self.res2.deviance_p)`

Let me know if there's anything else you want/need with this.

josef-pkt commented:
We need to rewrite and add to the Notes section in the docstring "Using frequency weights:"

"Using variance weights" needs a warning that loglikelihood/llf is not valid in models where scale is equal to 1, i.e. Binomial, Poisson and NegativeBinomial. If var_weights are specified, then some of the results like `llf` are based on a quasi-likelihood interpretation. The loglikelihood is not correctly specified in this case, and statistics based on it, such AIC or likelihood ratio tests, are not appropriate. This is similar to the case with robust covariances and the assumptions in GEE.

(Once the weight types have settled down and are extended to more models, then we need a documentation about what "weights in statsmodels" are, in comparison or contrast to R and Stata.)

josef-pkt commented:
np.asarray is also missing for `freq_weights`
I got an exception trying an example where freq_weights was a list.

josef-pkt commented:
I'm trying a very simple example with poisson and I'm getting confused again what should match theoretically when we aggregate or average. 
llf, deviance and pearson_chi2 seem not to be invariant in poisson.
I ran into the same confusion before when I tried log-linear models with GLM-Poisson. (If we aggregate with categorical explanatory variables we can end up with a saturated model with pearson_chi2=0 because we lost all dispersion information when we compute the sufficient statistics.)

(My guess is again that it is a theoretical issue and not computational, i.e. sufficient statistics are only sufficient for params, but not for the rest. There is too much literature about sufficient statistics which I never completely understood, and which look like it's overselling it for many use cases.)

thequackdaddy commented:
I took a stab at writing the notes for weights from my memory/notes. Probably a typo in there... I used your language/comments there to keep in clear. Let me know if you have any other ideas or want something else. 

I experimented with rebasing #3543 on master, and then rebasing this on said branch, and solving the merge conflicts. I think its easy enough. With your permission, I can either do that or wait for #3543 to be rebased and merged by its author.

thequackdaddy commented:
Sorry for all the pushes. I tried to write some tests to make sure that the warnings were appropriately raised, but they don't work on older pythons, so I skipped them. I'm not sure why this is, but I would guess buried deep in the code somewhere we silence some warnings and that is problematic on older pythons. 

Done for today, in all likelihood.

josef-pkt commented:
aside: I was trying to figure out how to aggregate or average with pandas for unique exog
(given that I have not much experience with pandas there might be better ways)
in this example:  
original shape (6366, 9)  
with freq_weights 684
with aggregation/averaging 130 (unique exog)

this seems to work to get mean and sum of endog for unique exog

```
In [75]:

gr = data['affairs rate_marriage age yrs_married'.split()].groupby('rate_marriage age yrs_married'.split())
df_a = gr.agg(['mean', 'sum','count'])
def merge_tuple(tpl):
    if isinstance(tpl, tuple) and len(tpl) > 1:
        return "_".join(map(str, tpl))
    else:
        return tpl
    
df_a.columns = df_a.columns.map(merge_tuple)
df_a.reset_index(inplace=True)
df_a.head()
Out[75]:
rate_marriage	age	yrs_married	affairs_mean	affairs_sum	affairs_count
0	1.0	17.5	0.5	0.000000	0.000000	1
1	1.0	22.0	2.5	3.643999	36.439994	10
2	1.0	27.0	2.5	3.200000	15.999998	5
3	1.0	27.0	6.0	0.660000	6.599998	10
4	1.0	27.0	9.0	0.855555	2.566665	3
```

and this for freq_weights for unique (endog, exog) rows

```

In [91]:
data2 = data.copy()
data2['const'] = 1
dc = data2['affairs rate_marriage age yrs_married const'.split()].groupby('affairs rate_marriage age yrs_married'.split()).count()
dc.reset_index(inplace=True)
dc.rename(columns={'const': 'freq'}, inplace=True)
print(dc.shape)
dc.head()

Out[91]:
(684, 5)

affairs	rate_marriage	age	yrs_married	freq
0	0.0	1.0	17.5	0.5	1
1	0.0	1.0	22.0	2.5	3
2	0.0	1.0	27.0	2.5	1
3	0.0	1.0	27.0	6.0	5
4	0.0	1.0	27.0	9.0	1
```

josef-pkt commented:
I was playing for a while with Poisson and Binomial examples.

I think the definitions in this PR are fine for now. (*)
Only combining identical observations and using freq_weights is lossless. Aggregation and averaging removes some information, that at least currently we don't recover. Getting `exposure` to work might be interesting because it doesn't have an ambiguous QMLE definition of `var_weights`.

(*) One case I didn't figure out yet is that I though we can replicate exposure with inverted var_weights. While parameters are just transformed, I didn't figure why standard errors differ.

I will try to go back tomorrow to reviewing the PRs so we can start to merge.
#3543 is overall very good, but still needs checking in some details.
One problem is that we are designing some parts for potential future use cases that we don't have yet, so it takes some guessing.

josef-pkt commented:
another piece of the puzzle:
using freq_weights or var_weights/average have the same llf and deviance difference when dropping one variable. So the difference is in the constant term that we don't correct, and it will not influence LR type tests within the same original/average data aggregation.

pearson_chi2 doesn't make sense in my poisson example right now because it has the wrong sign when dropping a variable when using exposure or var_weights with aggregated data. (no clue yet)

josef-pkt commented:
I'm not quite why. I didn't see this before, but pattern is similar to f-test based on ssr

If I use a common variance when computing pearson chi2, then I get identical results across aggregation patterns when dropping a variable. (notebooks are not nice for copy-pasting results)

res name with 2 at end has one fewer explanatory variables
o: original
f: with freq_weights
e: summed with exposure
a: averaged with var_weights

In this case I used denominator under the reduced/restricted model. i.e. similar to assuming variance is under the Null (extra coefficient=0)

```
In [175]:
((res_e2._results.resid_response**2 - res_e._results.resid_response**2) / res_e2.model.family.variance(res_e2.mu)).sum()
Out[175]:
27.195307546048578

In [185]:
((res_a2._results.resid_response**2 - res_a._results.resid_response**2) / res_a2.model.family.variance(res_a2.mu)
* res_a2.model.var_weights).sum()
Out[185]:
27.195307546048536

In [183]:
((res_f2._results.resid_response**2 - res_f._results.resid_response**2) / res_f2.model.family.variance(res_f2.mu) 
 * res_f2.model.freq_weights).sum()
Out[183]:
27.19530754604785

In [178]:
((res_o2._results.resid_response**2 - res_o._results.resid_response**2) / res_o2.model.family.variance(res_o2.mu)).sum()
Out[178]:
27.195307546049946
```

aside: example is the affairs dataset, where endog is float and not int, i.e. decimal values, so not strictly correct for Poisson.

josef-pkt commented:
I cleaned up a notebook at https://gist.github.com/josef-pkt/3a9a493a96ecc4617ab4a13dc5ce94e5
Essentially, the same type of models as before but uses aggregation of observations in the affairs data so we don't have to create artificial observations.

thequackdaddy commented:
Very good resource. Thanks for sharing that. 

I went ahead and updated the notebook in the PR with your changes. 

Let me know if you need anything else. Again, I'm happy (and relatively comfortable) with rebasing this #3543 on master and then rebasing this on that new branch, solving the merge conflicts, and making a new PR to get all of this.

Thanks, Josef.

josef-pkt commented:
I will go back to it on Sundary. No time until then.

I didn't see any new problems, so we just leave llf, deviance and pearson_chi2 as open issue.
I don't remember the details and references for example for #3101 well enough to figure out now what the proper behavior for "weird" cases should be.

josef-pkt commented:
(just a note while trying to get an overview of proportions again)
GLM binomial with counts (2-D endog) converts internally to proportion, and fittedvalues are proportions.  
So there should be an equivalence: using endog=proportion plus var_weights should be equivalent to 2-dim endog counts in the way glm binomial is implemented.  
And most likely 2-dim endog with proportions of (success, fail) plus var_weights for number of trials is also equivalent. If it works this would test that internal binomial weights and var_weights are correctly combined.
(I'm not sure we already looked at those cases.)

thequackdaddy commented:
> (just a note while trying to get an overview of proportions again)
GLM binomial with counts (2-D endog) converts internally to proportion, and fittedvalues are proportions.
So there should be an equivalence: using endog=proportion plus var_weights should be equivalent to 2-dim endog counts in the way glm binomial is implemented.
And most likely 2-dim endog with proportions of (success, fail) plus var_weights for number of trials is also equivalent. If it works this would test that internal binomial weights and var_weights are correctly combined.
(I'm not sure we already looked at those cases.)

I wrote a test to specifically test this. Looks like loglikelihood and deviance are too different in the formula interpretaiton between 1d and 2d for this to work. But `params` and `bse` seem to match.

josef-pkt commented:
> But params and bse seem to match.

I think we should merge #3543 and this PR fast with the warnings that we already have and then fiddle with some details. It will be easier to tell what's missing if both are merged. (*)

I'm still not sure about loglike and llf, it's difficult for me to figure out what we really want when we don't have full QMLE overdispersion support yet. (There are still bugs with loglike in this PR but maybe fewer in combination with #3543, families loglike will have to treat var_weights different from the other weights.)

(*) Another testcase where loglike/llf will fail is comparing GLM Gaussian and var_weights and WLS, based on a missing term in this PR in the families.Gaussian.loglike, AFAICS. This might or might not be fixed in 3543.

(I am also slow in adjusting to think about Binomial as a proportion endog instead of a count distribution.)
